### ISSUE TITLE
Are NA values allowed in 'y' for supervised reduction?

### ISSUE THREAD

**Issue by ummel** [jlmelville/uwot#9]
UMAP supports partial labeling (i.e. NA values) of a target array when performing supervised reduction.

And this is what `uwot::umap()` does when the the 'y' argument is a character:

```
x <- iris
x$Species[sample.int(nrow(iris), 50)] <- NA
iris_umap <- umap(x[,-5], n_neighbors = 50, alpha = 0.5, init = "random", y = x$Species)
```
However, this fails in the case of a _numeric_ 'y' argument that contains NA values:

```
x <- mtcars
x$mpg[sample.int(nrow(mtcars), 10)] <- NA
mtcars_umap <- umap(x[,-1], n_neighbors = 10, alpha = 0.5, init = "random", y = x$mpg)
```
> Error in result[n_samples > 0] <- n_epochs/n_samples[n_samples > 0] : 
  NAs are not allowed in subscripted assignments

Perhaps this is the expected behavior (I admit I am not familiar with the details of the algorithm), but I wanted to confirm if this is the case or not.

**Comment by jlmelville** [comment]
Yes, I'm afraid one place where uwot lags behind the Python implementation is in the choice of distance metric. Only those supported by RcppAnnoy are allowed.

The numeric values of `y` are used to generate nearest neighbor data, so if you had a way of converting `y` into such data externally to `uwot::umap`, I could modify the routine to allow the `y` parameter to accept nearest neighbor info. That's certainly something I would consider if it would be useful. 

The input format would need to be the same as that expected by the `nn_method` parameter when passing nearest neighbor info directly: a list consisting of two matrices, both with dimension n_vertices x n_neighbors: `idx`, containing the integer indexes of the nearest neighbors, and `dist`, containing the distances.

**Comment by ummel** [comment]
So I understand: You're proposing a `y`-specific argument of flavor `nn_method` that could "manually" specify the nearest neighbors and distance between `y` and `X`?

If so, that could be interesting. At the very least, it would give users full(er) control for addressing or testing unusual issues with input data.

**Comment by jlmelville** [comment]
When `y` is numeric it goes through a very similar transformation as `X` does: nearest neighbors are calculated (via Euclidean distance) and then a fuzzy simplicial set is created. For supervised UMAP, the intersection of the `X` and `y` sets is then used in the main UMAP algorithm.

So my proposal is: instead of passing the numeric vector to `y`, pass the nearest neighbor data generated from the numeric vector instead. I can add support for this fairly easily.

**Comment by jlmelville** [comment]
I have added a section to `README.md` on the nearest neighbor data format expected by `y` (or `nn_method` for that matter). Not sure there is much else I can add for now. I will probably close this issue when I create the next release.
